README.md

```diff
@@ -5,4 +5,402 @@
+    if ok, val := failpoint.Eval("failpoint-name-2"); ok {
+        return rand.Intn(val.(int))
+    }
+    return rand.Intn(6000)
+}() {
+    a()
+    if ok, val := failpoint.Eval("failpoint-name-3"); ok {
+        fmt.Println("unit-test", val)
+    }
+}
+```
+
+- Inject failpoint to the SELECT statement to make it block one case if failpoint is active
+
+```go
+func (s *StoreService) ExecuteStoreTask() {
+    select {
+    case <-func() chan *StoreTask {
+        failpoint.Inject("priority-fp", func(_ failpoint.Value) {
+            return make(chan *StoreTask)
+        })
+        return s.priorityHighCh
+    }():
+        fmt.Println("execute high priority task")
+
+    case <- s.priorityNormalCh:
+        fmt.Println("execute normal priority task")
+
+    case <- s.priorityLowCh:
+        fmt.Println("execute normal low task")
+    }
+}
+```
+
+Will generate something like this:
+
+```go
+func (s *StoreService) ExecuteStoreTask() {
+    select {
+    case <-func() chan *StoreTask {
+        if ok, _ := failpoint.Eval("priority-fp"); ok {
+            return make(chan *StoreTask)
+        })
+        return s.priorityHighCh
+    }():
+        fmt.Println("execute high priority task")
+
+    case <- s.priorityNormalCh:
+        fmt.Println("execute normal priority task")
+
+    case <- s.priorityLowCh:
+        fmt.Println("execute normal low task")
+    }
+}
+```
+
+- Inject a failpoint to dynamically extend SWITCH CASE arms
+
+```go
+switch opType := operator.Type(); {
+case opType == "balance-leader":
+    fmt.Println("create balance leader steps")
+
+case opType == "balance-region":
+    fmt.Println("create balance region steps")
+
+case opType == "scatter-region":
+    fmt.Println("create scatter region steps")
+
+case func() bool {
+    failpoint.Inject("dynamic-op-type", func(val failpoint.Value) bool {
+        return strings.Contains(val.(string), opType)
+    })
+    return false
+}():
+    fmt.Println("do something")
+
+default:
+    panic("unsupported operator type")
+}
+```
+
+Will generate something like this:
+
+```go
+switch opType := operator.Type(); {
+case opType == "balance-leader":
+    fmt.Println("create balance leader steps")
+
+case opType == "balance-region":
+    fmt.Println("create balance region steps")
+
+case opType == "scatter-region":
+    fmt.Println("create scatter region steps")
+
+case func() bool {
+    if ok, val := failpoint.Eval"dynamic-op-type"); ok {
+        return strings.Contains(val.(string), opType)
+    })
+    return false
+}():
+    fmt.Println("do something")
+
+default:
+    panic("unsupported operator type")
+}
+```
+
+- More complicated failpoint
+
+    - There are more complicated failpoint sites that can be injected to
+        - for the loop INITIAL statement, CONDITIONAL expression and POST statement
+        - for the RANGE statement
+        - SWITCH INITIAL statement
+        - …
+    - Anywhere you can write a function literal
+
+## Implementation details
+
```

Comment:
If the following shows the implementation steps in order, it is better to use an ordered list (1, 2, 3, 4) here.